v2/pkg/engine/resolve/loader.go

```diff
@@ -412,31 +412,45 @@ func (l *Loader) mergeResult(res *result, items []int) error {
 		return nil
 	}
 	if res.out.Len() == 0 {
-		return l.renderErrorsFailedToFetch(res, failedToFetchEmptyResponse)
+		return l.renderErrorsFailedToFetch(res, emptyGraphQLResponse)
 	}
-	node, err := l.data.AppendAnyJSONBytes(res.out.Bytes())
+
+	// We expect a JSON object according to GraphQL over HTTP spec
+	node, err := l.data.AppendObject(res.out.Bytes())
 	if err != nil {
-		return l.renderErrorsFailedToFetch(res, failedToFetchInvalidJSON)
+		return l.renderErrorsFailedToFetch(res, invalidGraphQLResponse)
 	}
 
-	// error handling
+	hasErrors := false
+
+	// We check if the subgraph response has errors
 	if res.postProcessing.SelectResponseErrorsPath != nil {
-		// look for errors in the response and merge them into the errors array
 		ref := l.data.Get(node, res.postProcessing.SelectResponseErrorsPath)
 		if ref != -1 {
+			hasErrors = l.data.NodeIsDefined(ref) && len(l.data.Nodes[ref].ArrayValues) > 0
+			// Look for errors in the response and merge them into the errors array
 			err = l.mergeErrors(res, ref)
 			if err != nil {
 				return errors.WithStack(err)
 			}
 		}
 	}
+
+	// We also check if any data is there to processed
 	if res.postProcessing.SelectResponseDataPath != nil {
 		node = l.data.Get(node, res.postProcessing.SelectResponseDataPath)
+		// Check if the data is an array and not null
 		if !l.data.NodeIsDefined(node) {
```

Comment:
Did you miss to check if data is an object?

Reply:
Updated. It seems in some scenarios we also work with arrays. I moved the validation to the `SelectResponseDataPath` path.